### DISCUSSION TITLE
Is it possible to ignore the ESC key if the IDE autocomplete window is open?

### DISCUSSION THREAD

**Discussion by rickgladwin** [JetBrains/ideavim#687] [Q&A]
Here's the use case that's affecting my productivity:
- enter input mode
- begin typing
- IDE autocomplete window appears (TAB or Enter will accept the autocomplete suggestion, ESC will dismiss the window)
- press ESC
- autocomplete window disappears _and_ Vim exits insert mode
- enter insert mode again
- (since the autocomplete window has been recently dismissed, further typing does not immediately relaunch the window)

I asked about this over on the JetBrains forum as well:
https://intellij-support.jetbrains.com/hc/en-us/community/posts/115000679170/comments/12861128642322

Surely this is a common issue when using both IdeaVim and the IDE native autocomplete. Has anyone found a solution?

Is there a way to detect when the autocomplete window is active, and ignore the ESC key when that's the case? I imagine this could be done with a vim script, but we'll need something on which to test the conditional.

Another option is to change the vim keymapping to require two ESC presses, but that runs counter to the reason I use vim, namely to _reduce_ time and keypresses.

Another option is to use vim plugins for autocomplete (if that's possible) and disable the native autocomplete, but I'd lose the rest of the autocomplete functionality (TabNine, GitHub Copilot, other IDE features)

Cheers

**Comment by AlexPl292** [comment]
Hi there. This is the original behavior of Vim (both closing completion and exiting insert mode). Vim does have to commands to close the completion without exiting insert mode, but they're currently not supported by IdeaVim: [VIM-3049](https://youtrack.jetbrains.com/issue/VIM-3049)

**Comment by rickgladwin** [comment]
Thanks very much for the link, @AlexPl292 

I think it's a strange default behaviour (we're using a single keypress to trigger two different actions in two different features, one of which has been activated automatically), but there seems to be consensus in those other threads and tickets that that's the expected behaviour, and it does keep the result of the keypress state-independent.

Enabling `Settings | Appearance & Behavior | Appearance | Support screen readers` per [VIM-2772...screen reader workaround](https://youtrack.jetbrains.com/issue/VIM-2772/Regression-of-Escape-key-in-insert-mode-closes-autocomplete-dropdown-without-exiting-insert-mode#focus=Comments-27-7055054.0-0) actually results in the behaviour I'm looking for in PyCharm 2023.2 (Professional Edition) - Build #PY-232.8660.197

Any hints where I should look around in https://github.com/JetBrains/ideavim to see if I can find the relationship between the `Support screen readers` setting and the ESC behaviour? I don't expect the workaround to last forever. Thanks!

**Reply by AlexPl292** [reply]
Also, it is worth mentioning that some time ago, we accidentally changed this behavior and it resulted in a lot of disappointed responses: [IDEA-209779](https://youtrack.jetbrains.com/issue/IDEA-209779).

The thing with a screen reader is not a workaround but rather a bug, and we are currently unaware of how this feature affects IdeaVim internally.

**Reply by huyz** [reply]
You probably don't to enable Screen readers if you want to see tooltips in the new IntelliJ UI (another quirk of the "Support screen readers" feature)

**Comment by huyz** [comment]
Yeah, we're in bad need of a separate command like `hideSuggestWidget` in VS Code.

**Comment by vladbat00** [comment]
To anyone who's still looking for a workaround: `imap .. <Action>(EditorEscape)`
